### ISSUE TITLE
protocol documentation?

### ISSUE THREAD

**Issue by protobits** [JiriBilek/WiFiSpiESP#14]
In case I want to communicate with the module from an RTOS (not Arduino), I would need the protocol definition. Is it planned?

**Comment by jandrassy** [comment]
1) what module?
2) this is an Arduino library
3) what protocol?

**Comment by JiriBilek** [comment]
Do you want to make yourself the other side (master side) of communication?
In that case you can obviously use the Arduino project (WifiSPI) as a start. I think you'll need to change only the lowest layer of communication.
There is no magic in the protocol but it would require some tome do write a documentation.

**Comment by protobits** [comment]
I would prefer to have a documentation instead of an Arduino library since I was interested in porting to NuttX RTOS. Is the protocol mostly stable at this point? If so, I could try to infer it from the code. In any case, any documentation you can provide would be great.

**Comment by endlos99** [comment]
I agree with v01d here.  I'm trying to connect an STM32 as SPI master to an ESP using this library, and it doesn't work at all.

The STM32 allows to configure the SPI master with various parameters:
* Data size
* Clock polarity (CPOL): High or low
* Clock phase (CPHA): 1st edge or 2nd edge
* NSS pulse mode: On or off

What parameters is the library implementing?  How many bytes are transferred at the minimum (the original Arduino ESP library uses 32 bytes)?

**Comment by JiriBilek** [comment]
Which framework are you using with the STM32? I tested the library with [https://github.com/rogerclarkmelbourne/Arduino_STM32](https://github.com/rogerclarkmelbourne/Arduino_STM32) and it worked well.

**Comment by endlos99** [comment]
I'm using the HAL.

**Comment by JiriBilek** [comment]
Unfortunately, I don't have all the data here on this computer but I think, the setup is:
CPOL=0, CPHA=0, i.e. SPI_MODE = 0.
You need to set the SS to zero before the transmission and pull it back to 1 afterwards.
There could be one more catch - place a small delay (try 30 ms) between SS = 0 and start of the transmission. The ESP8266 SPI implementation is far from ideal.
You can see [here](https://user-images.githubusercontent.com/15474360/49903651-2ecc6080-fe68-11e8-9b1d-684de09ba4f5.png) for a logic analyzer picture. See the lower picture.

**Comment by endlos99** [comment]
Unfortunately, what you describe (SS low before transmission, high after) is CPOL=1, since CPOL defines the idle state.  CPOL=1 would also match the comment in the Arduino library (high = deselected) and the fact that you set SPI1C2 to sample on rising edge.

**Comment by endlos99** [comment]
Sorry, never mind, I think I mixed up CLK and SS in above comment.

**Comment by JiriBilek** [comment]
I dug into the master library and the setting is:
MSBFIRST, SPI_MODE0.

For STM32F103 it is: 
SPI_CR1.LSBFIRST = 0
SPI_CR1.CPOL = 0
SPI_CR1.CPHA = 0

**Comment by endlos99** [comment]
Thanks, that helps!  But what about the data size?  The STM32 I'm using only supports a data size of 16 bit.  I couldn't really find a data size configuration for the library, only the buffer size.

With my current configuration, I receive data at the slave, but the bytes are entirely corrupted.

Do I need to make a change to the library to support 16 bits?  (The IR handler for case status & SPISWBIS comes to mind.)

**Comment by JiriBilek** [comment]
The datagram format is not arbitrary. The ESP8266 SPI slave can work with data of two lengths only - 4 bytes and 32 bytes. The library uses the smaller size for status messages and the bigger one for data messages. It's the limitation of the hardware.

**Comment by endlos99** [comment]
OK, got it, thanks!
